File: src/sales_formatter.py
"""
sales_formatter.py — Formats the Sales follow-up email for rep review.

The script sends an email TO the rep (not the prospect) containing a
pre-formatted draft the rep can review, edit, and send from their own account.

Prospect-facing template:
  Subject: SiteZeus Resources & Next Steps
  Hi {First Name},
  Thank you for taking the time to connect with me today...
  What we discussed: ... (from Gong brief or key points)
  Next steps: ... (from Gong actions or transcript)
  Resources: ...
  Best, {Rep Name}
"""
import logging
from typing import List, Optional, Tuple

import config

logger = logging.getLogger(__name__)

# Highlight types to EXCLUDE from the "Quick highlights" section
EXCLUDE_HIGHLIGHT_TYPES = {
    "action_item", "action item", "next_step", "next step",
    "decision", "key_decision",
}

NEXT_MEETING_TYPES = {"next_meeting", "follow_up", "follow-up", "upcoming_meeting"}


def build_prospect_html(
    call_data: dict,
    rep_name: str,
    prospect_first_name: str,
    is_enterprise: bool,
) -> str:
    """Build just the prospect-facing email body (used for dashboard preview)."""
    content = call_data.get("content", {})
    highlights = content.get("highlights", [])
    trackers = content.get("trackers", [])
    call_highlights = _extract_highlights(highlights)[:5]
    tracker_names = [t.get("name", "").lower() for t in trackers if (t.get("count") or 0) > 0]
    resources = _select_resources(tracker_names, is_enterprise)
    next_meeting = _detect_next_meeting(highlights)
    return _build_prospect_html(
        prospect_first_name=prospect_first_name,
        rep_name=rep_name,
        highlights=call_highlights,
        resources=resources,
        next_meeting=next_meeting,
    )


def format_sales_review_email(
    call_data: dict,
    rep_name: str,
    rep_email: str,
    prospect_first_name: str,
    prospect_email: str,
    is_enterprise: bool,
    call_name: str,
    transcript: list = None,
) -> Tuple[str, str, str]:
    """
    Build the sales review email sent to the rep.
    Returns (subject, wrapper_html, prospect_preview_html).
    """
    content = call_data.get("content", {}) or {}
    highlights = content.get("highlights", []) or []
    trackers = content.get("trackers", []) or []

    # ----------------------------------------------------------------
    # 1. "What we discussed" — prefer Gong AI brief > key points > highlights > transcript
    # ----------------------------------------------------------------
    brief_text = _clean_text(content.get("brief"))
    key_points = _extract_key_points(content.get("keyPoints") or content.get("key_points"))
    call_highlights = _extract_highlights(highlights)[:5]

    # If Gong AI returned no highlights, extract key sentences from transcript
    if not call_highlights and transcript:
        call_highlights = _extract_from_transcript(transcript, rep_name)[:5]

    logger.info(
        "Gong AI fields — brief: %s, keyPoints: %d, highlights: %d",
        "yes" if brief_text else "no",
        len(key_points),
        len(call_highlights),
    )

    # ----------------------------------------------------------------
    # 2. Resources — from trackers, then outline/topics, then transcript
    # ----------------------------------------------------------------
    tracker_names = [
        t.get("name", "").lower()
        for t in trackers
        if (t.get("count") or 0) > 0
    ]
    # Try outline/topics fields for topic signals
    outline_topics = _extract_outline_topics(content.get("outline") or content.get("topics"))
    tracker_names = list(set(tracker_names + outline_topics))

    # Also scan transcript for product mentions if still empty
    if not tracker_names and transcript:
        tracker_names = _detect_topics_from_transcript(transcript)

    resources = _select_resources(tracker_names, is_enterprise)

    # ----------------------------------------------------------------
    # 3. Next steps — prefer Gong actions > action_item highlights > transcript
    # ----------------------------------------------------------------
    next_steps = _extract_gong_actions(content.get("actions"))
    if not next_steps:
        next_steps = [
            (h.get("text") or "").strip()
            for h in highlights
            if h.get("type", "").lower() in ("action_item", "action item", "next_step", "next step")
            and (h.get("text") or "").strip()
        ]
    if not next_steps and transcript:
        next_steps = _extract_action_items_from_transcript(transcript)[:5]

    # Detect next agreed meeting
    next_meeting = _detect_next_meeting(highlights)
    if not next_meeting and transcript:
        next_meeting = _detect_next_meeting_from_transcript(transcript)

    prospect_html = _build_prospect_html(
        prospect_first_name=prospect_first_name,
        rep_name=rep_name,
        brief_text=brief_text,
        key_points=key_points,
        highlights=call_highlights,
        resources=resources,
        next_meeting=next_meeting,
        next_steps=next_steps,
    )

    segment_label = "Enterprise" if is_enterprise else "SMB / Emerging"
    wrapper_html = _build_rep_wrapper(
        rep_email=rep_email,
        prospect_first_name=prospect_first_name,
        prospect_email=prospect_email,
        segment_label=segment_label,
        call_name=call_name,
        prospect_html=prospect_html,
    )

    subject = f"[REVIEW & SEND] Follow-up draft for {prospect_first_name} — {call_name}"
    return subject, wrapper_html, prospect_html


# ------------------------------------------------------------------
# Internal helpers
# ------------------------------------------------------------------

def _clean_text(val) -> Optional[str]:
    """Return stripped non-empty string or None."""
    if not val:
        return None
    if isinstance(val, str):
        t = val.strip()
        return t if t else None
    return None


def _extract_key_points(key_points_field) -> List[str]:
    """Extract bullet points from Gong keyPoints field (list of dicts or strings)."""
    if not key_points_field:
        return []
    items = []
    if isinstance(key_points_field, list):
        for kp in key_points_field:
            if isinstance(kp, str):
                t = kp.strip()
                if t:
                    items.append(t)
            elif isinstance(kp, dict):
                t = _clean_text(kp.get("text") or kp.get("title") or kp.get("description"))
                if t:
                    items.append(t)
    return items[:6]


def _extract_gong_actions(actions_field) -> List[str]:
    """Extract next steps from Gong actions field."""
    if not actions_field:
        return []
    items = []
    if isinstance(actions_field, list):
        for a in actions_field:
            if isinstance(a, str):
                t = a.strip()
                if t:
                    items.append(t)
            elif isinstance(a, dict):
                t = _clean_text(a.get("text") or a.get("description") or a.get("title"))
                if t:
                    items.append(t)
    return items[:5]


def _extract_outline_topics(outline_field) -> List[str]:
    """Extract topic names from Gong outline or topics field for resource matching."""
    if not outline_field:
        return []
    topics = []
    if isinstance(outline_field, list):
        for item in outline_field:
            if isinstance(item, str):
                topics.append(item.lower())
            elif isinstance(item, dict):
                name = item.get("name") or item.get("title") or item.get("topic") or ""
                if name:
                    topics.append(name.lower())
    elif isinstance(outline_field, str):
        topics.append(outline_field.lower())
    return topics


def _extract_highlights(highlights: list) -> List[str]:
    texts = []
    for h in highlights:
        h_type = (h.get("type") or "").lower()
        text = (h.get("text") or "").strip()
        if text and h_type not in EXCLUDE_HIGHLIGHT_TYPES:
            texts.append(text)
    return texts


def _select_resources(tracker_names: List[str], is_enterprise: bool) -> List[dict]:
    """
    Pick resources whose topics overlap with what was discussed on the call.
    Fill remaining slots from the top of the resource list.
    Always appends a Customer Stories link at the end.
    Max 4 resources + Customer Stories.
    """
    pool = config.ENTERPRISE_RESOURCES if is_enterprise else config.SMB_EMERGING_RESOURCES
    selected = []
    selected_names = set()

    # Topic-matched resources first
    for resource in pool:
        if len(selected) >= 4:
            break
        for topic in resource.get("topics", []):
            if any(topic in tracker for tracker in tracker_names):
                if resource["name"] not in selected_names:
                    selected.append(resource)
                    selected_names.add(resource["name"])
                break

    # Fill remaining slots from pool top
    for resource in pool:
        if len(selected) >= 4:
            break
        if resource["name"] not in selected_names:
            selected.append(resource)
            selected_names.add(resource["name"])

    # Always include customer stories
    selected.append({"name": "Customer Stories", "url": config.CUSTOMER_STORIES_URL})
    return selected


def _detect_next_meeting(highlights: list) -> Optional[str]:
    for h in highlights:
        h_type = (h.get("type") or "").lower()
        text = (h.get("text") or "").strip()
        if h_type in NEXT_MEETING_TYPES and text:
            return text
    return None


def _bullet_list(items) -> str:
    return "<ul>" + "".join(f"<li>{i}</li>" for i in items) + "</ul>"


def _build_prospect_html(
    prospect_first_name: str,
    rep_name: str,
    highlights: List[str],
    resources: List[dict],
    next_meeting: Optional[str],
    next_steps: List[str] = None,
    brief_text: Optional[str] = None,
    key_points: List[str] = None,
) -> str:
    parts = []
    parts.append(f"<p>Hi {prospect_first_name},</p>")
    parts.append(
        "<p>Thank you for taking the time to connect with me today. "
        "I wanted to follow up with a quick recap of what we covered.</p>"
    )

    # ---- What we discussed ----
    # Priority: Gong AI brief (paragraph) > key points (bullets) > highlights (bullets)
    if brief_text:
        parts.append("<p><strong>What we discussed:</strong></p>")
        # Render brief as a paragraph — it's already a clean AI summary
        parts.append(f"<p>{brief_text}</p>")
    elif key_points:
        parts.append("<p><strong>What we discussed:</strong></p>")
        parts.append(_bullet_list(key_points))
    elif highlights:
        parts.append("<p><strong>What we discussed:</strong></p>")
        parts.append(_bullet_list(highlights))

    # ---- Next steps ----
    if next_steps:
        parts.append("<p><strong>Next steps:</strong></p>")
        parts.append(_bullet_list(next_steps))
    elif next_meeting:
        parts.append("<p><strong>Next steps:</strong></p>")
        parts.append(f"<p>{next_meeting}</p>")
    else:
        parts.append("<p><strong>Next steps:</strong></p>")
        parts.append(
            "<p>I'll be in touch soon. Feel free to reach out with any questions in the meantime. "
            "You can find time through the scheduling link in my signature.</p>"
        )

    # ---- Resources ----
    if resources:
        parts.append("<p><strong>Resources:</strong></p>")
        resource_items = [
            f'<a href="{r["url"]}" style="color:#1a73e8;">{r["name"]}</a>'
            for r in resources
        ]
        parts.append(_bullet_list(resource_items))

    parts.append(f"<p>Best,<br>{rep_name}</p>")
    return "\n".join(parts)


PRODUCT_KEYWORDS = {
    "locate": ["locate", "site selection", "forecast", "revenue", "model", "prediction"],
    "build": ["build", "construction", "milestone", "opening", "timeline", "project"],
    "sell": ["sell", "franchise", "crm", "pipeline", "candidate", "territory"],
    "market": ["market", "customer insights", "trade area", "mobile data", "consumer"],
    "zeus_ai": ["zeus", "ai", "artificial intelligence", "predict", "recommendation"],
    "white_space": ["white space", "whitespace", "expansion", "growth", "new market"],
    "sales_impact": ["cannibalization", "impact", "transfer", "overlap"],
    "poc": ["proof of concept", "poc", "pilot", "trial", "test"],
    "pricing": ["pricing", "price", "cost", "investment", "contract", "proposal"],
}

NEXT_MEETING_KEYWORDS = ["next week", "follow up", "follow-up", "schedule", "tuesday",
                          "wednesday", "thursday", "monday", "friday", "next call",
                          "next meeting", "demo", "presentation"]


def _extract_from_transcript(transcript: list, rep_name: str) -> List[str]:
    """Extract meaningful sentences from transcript as highlights."""
    rep_first = rep_name.split()[0].lower() if rep_name else ""
    highlights = []
    seen = set()

    for s in transcript:
        text = (s.get("text") or "").strip()
        if len(text) < 20 or len(text) > 200:
            continue
        tl = text.lower()
        # Skip small talk and generic openers
        if any(skip in tl for skip in ["how are you", "happy friday", "nice to meet",
                                        "how's it going", "sounds good", "absolutely",
                                        "yeah", "okay", "alright", "perfect"]):
            continue
        # Prefer sentences that mention SiteZeus products or business value
        for group, keywords in PRODUCT_KEYWORDS.items():
            if any(kw in tl for kw in keywords) and text not in seen:
                highlights.append(text)
                seen.add(text)
                break
        if len(highlights) >= 5:
            break

    return highlights


def _detect_topics_from_transcript(transcript: list) -> List[str]:
    """Scan transcript for product/topic mentions to guide resource selection."""
    topics = set()
    full_text = " ".join((s.get("text") or "").lower() for s in transcript)
    for topic, keywords in PRODUCT_KEYWORDS.items():
        if any(kw in full_text for kw in keywords):
            topics.add(topic.replace("_", " "))
    return list(topics)


def _extract_action_items_from_transcript(transcript: list) -> List[str]:
    """Extract action items — sentences where someone commits to doing something."""
    action_phrases = [
        "i will", "i'll", "we will", "we'll", "i'm going to",
        "will send", "will follow", "will check", "will get back",
        "will discuss", "will review", "will share", "will reach out",
        "will connect", "by end of", "by next week", "by friday",
        "going to send", "going to follow"
    ]
    items = []
    seen = set()
    for s in transcript:
        text = (s.get("text") or "").strip()
        tl = text.lower()
        if 15 < len(text) < 200 and any(p in tl for p in action_phrases):
            if text not in seen:
                items.append(text)
                seen.add(text)
    return items


def _detect_next_meeting_from_transcript(transcript: list) -> Optional[str]:
    """Scan transcript for next meeting references."""
    for s in reversed(transcript):
        text = (s.get("text") or "").strip()
        tl = text.lower()
        if any(kw in tl for kw in NEXT_MEETING_KEYWORDS) and len(text) < 150:
            return text
    return None


def _build_rep_wrapper(
    rep_email: str,
    prospect_first_name: str,
    prospect_email: str,
    segment_label: str,
    call_name: str,
    prospect_html: str,
) -> str:
    return f"""
<div style="font-family:Arial,Helvetica,sans-serif;font-size:14px;color:#333;max-width:680px;">

  <div style="background:#fff8e1;padding:14px 18px;border-left:5px solid #f9a825;margin-bottom:20px;border-radius:3px;">
    <strong>ACTION REQUIRED — Review &amp; Send</strong><br><br>
    This is a draft follow-up email for <strong>{prospect_first_name}</strong>
    (<a href="mailto:{prospect_email}">{prospect_email}</a>).<br>
    <strong>Segment:</strong> {segment_label} &nbsp;|&nbsp;
    <strong>Call:</strong> {call_name}<br><br>
    Review the draft below. Edit as needed, then send it to the prospect from
    your own email account (<strong>{rep_email}</strong>).<br>
    <em>Do not reply to this message — it was sent automatically.</em>
  </div>

  <hr style="border:none;border-top:1px solid #ddd;margin:0 0 20px 0;">

  <p style="color:#888;font-size:12px;margin-bottom:4px;">
    Suggested subject: <strong>SiteZeus Resources &amp; Next Steps</strong>
  </p>

  <div style="border:1px solid #e0e0e0;padding:20px;border-radius:4px;background:#fafafa;">
    {prospect_html}
  </div>

</div>
"""

File: src/salesFormatter.js
// salesFormatter.js — Formats the Sales follow-up email for rep review.
//
// Claude reads the call transcript and writes a short, human-sounding
// follow-up from the perspective of the SiteZeus rep who led the call.
// Resources are selected from the curated catalog in config.js (the script
// has no live web access) and given a call-specific reason.

import * as config from "./config.js";
import { extractJson } from "./anthropic.js";

function escapeHtml(s) {
  return String(s || "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

// The template already opens with "Hi {name}," + "Thank you for hopping on the
// call today." Strip any leading greeting/pleasantry the model adds anyway, so
// the body starts with the recap rather than a second hello.
function stripLeadingGreeting(body) {
  if (!body) return body;
  let text = body.trimStart();
  const GREETING_RE = /^(hi|hey|hello|dear)\b[^.!?\n]*[.!?\n]+\s*/i;
  const PLEASANTRY_RE = /^[^.!?\n]*\b(really enjoyed|enjoyed (the |our )?(call|conversation|chat|connecting|meeting)|great (catching up|to (connect|chat|meet|speak)|connecting|meeting|chatting|speaking)|glad we (got to )?(connect|chat|met|spoke|connected|talked)|good (talking|speaking|connecting)|thanks (so much )?for|thank you (so much )?for|appreciate you|nice (to )?(meet|chat|connect|speak))\b[^.!?\n]*[.!?]+\s*/i;
  for (let i = 0; i < 2; i++) {
    if (GREETING_RE.test(text)) { text = text.replace(GREETING_RE, ""); continue; }
    if (PLEASANTRY_RE.test(text)) { text = text.replace(PLEASANTRY_RE, ""); continue; }
    break;
  }
  return text.trim() || body.trim(); // never return empty
}

function buildTranscriptText(transcript, parties) {
  if (!transcript?.length) return "";
  const speakerMap = {};
  for (const party of (parties || [])) {
    const id = party.speakerId || party.userId || party.id;
    if (!id) continue;
    const isInternal =
      party.affiliation === "internal" ||
      (party.emailAddress || "").toLowerCase().includes(config.SITEZEUS_DOMAIN || "sitezeus.com");
    const first = (party.name || (isInternal ? "Rep" : "Prospect")).split(" ")[0];
    speakerMap[id] = isInternal ? `${first} (SiteZeus)` : `${first} (Prospect)`;
  }
  const utterances = [];
  let current = null;
  for (const s of transcript) {
    if (!s.text?.trim()) continue;
    const name = speakerMap[s.speakerId] || "Speaker";
    if (!current || current.name !== name) {
      current = { name, texts: [] };
      utterances.push(current);
    }
    current.texts.push(s.text.trim());
  }
  return utterances.map(u => `${u.name}: ${u.texts.join(" ")}`).join("\n");
}

// Pick the SiteZeus participant who spoke the most — they led the call and
// the follow-up should appear to come from them.
function primaryInternalSpeaker(transcript, parties) {
  const internalNames = {};
  for (const p of (parties || [])) {
    const id = p.speakerId || p.userId || p.id;
    if (!id) continue;
    const isInternal =
      p.affiliation === "internal" ||
      (p.emailAddress || "").toLowerCase().includes(config.SITEZEUS_DOMAIN || "sitezeus.com");
    if (isInternal && p.name) internalNames[id] = p.name.trim();
  }
  const words = {};
  for (const s of (transcript || [])) {
    if (!internalNames[s.speakerId] || !s.text) continue;
    words[s.speakerId] = (words[s.speakerId] || 0) + s.text.trim().split(/\s+/).length;
  }
  let bestId = null, best = -1;
  for (const [id, n] of Object.entries(words)) if (n > best) { best = n; bestId = id; }
  if (!bestId) return null;
  const name = internalNames[bestId];
  return { name, firstName: name.split(" ")[0] };
}

// Unique resource catalog the model may choose from.
function resourceCatalog() {
  const seen = new Set();
  const list = [];
  for (const r of config.ENTERPRISE_RESOURCES) {
    if (seen.has(r.name)) continue;
    seen.add(r.name);
    list.push(r);
  }
  return list;
}

async function extractSalesContent({ transcriptText, prospectFirstName, prospectCompany, senderName }) {
  if (!transcriptText) return null;

  const companyRef = prospectCompany || `${prospectFirstName}'s company`;
  const catalog = resourceCatalog();
  const catalogText = catalog.map(r => `- ${r.name} — topics: ${(r.topics || []).join(", ")}`).join("\n");

  const prompt = `You are ${senderName}, a SiteZeus sales rep who was on this call with ${prospectFirstName} from ${companyRef}. You are drafting the BODY of a short, human follow-up email. The email is ALREADY opened for you with a greeting line ("Hi ${prospectFirstName},") and a second line ("Thank you for hopping on the call today.") — your text comes immediately AFTER those two lines.

Transcript:
${transcriptText.slice(0, 14000)}

Return ONLY this JSON — no markdown, no commentary:
{
  "body": "The greeting and thank-you line are ALREADY written — your body must NOT include any greeting, must NOT thank them for their time, and must NOT address them by name or say things like 'great catching up' / 'really enjoyed the conversation'. Start the FIRST sentence directly with the recap: e.g. 'We discussed your goal of...' or 'You shared that your team is looking for a more efficient way to...'. 1-2 short paragraphs (separate with a blank line) connecting the prospect's goal -> the challenge they described -> how SiteZeus helps. First person ('I', 'we', 'our team'), second person ('you', 'your'). Reference specific things from THIS call. Frame challenges positively. No buzzwords, no overselling, no invented capabilities.",
  "nextSteps": ["Each explicitly agreed next step, with date/time if it was stated, e.g. 'Demo scheduled for July 18 at 2:00 PM ET'"],
  "resources": [ { "name": "EXACT name from the list below", "reason": "one short sentence tying it to something specific discussed on this call" } ]
}

Available resources (choose 0-5, only the genuinely relevant ones; if none fit, return []):
${catalogText}

Rules:
- body: max ~150 words, outcome-focused, never restate the whole call, never invent anything.
- nextSteps: ONLY steps explicitly discussed. Empty array if none. Never invent.
- resources: pick from the list by EXACT name only. Each needs a personalized, call-specific reason. Never recommend generic content.
- Keep the whole thing under 250 words. Do not mention features that were not discussed.`;

  return extractJson({ prompt, maxTokens: 900, temperature: 0.5, label: "Sales" });
}

// Map AI-selected resource names back to catalog URLs, dropping any the model
// invented. Returns [{ name, url, reason }].
function resolveResources(aiResources) {
  const urlByName = {};
  for (const r of resourceCatalog()) urlByName[r.name.toLowerCase()] = r.url;
  const out = [];
  const seen = new Set();
  for (const r of (aiResources || [])) {
    const url = urlByName[(r.name || "").toLowerCase()];
    if (!url || seen.has(url)) continue;
    seen.add(url);
    out.push({ name: r.name, url, reason: (r.reason || "").trim() });
    if (out.length >= 5) break;
  }
  return out;
}

// Fallback resource selection by topic keyword (used when AI is unavailable).
function selectResourcesByTopic(topics, isEnterprise) {
  const pool = isEnterprise ? config.ENTERPRISE_RESOURCES : config.SMB_EMERGING_RESOURCES;
  const selected = [];
  const seen = new Set();
  for (const resource of pool) {
    if (selected.length >= 3) break;
    for (const topic of resource.topics || []) {
      if (topics.some(t => t.includes(topic) || topic.includes(t)) && !seen.has(resource.name)) {
        selected.push({ name: resource.name, url: resource.url, reason: "" });
        seen.add(resource.name);
        break;
      }
    }
  }
  return selected;
}

export async function formatSalesReviewEmail({ callData, repName, repEmail, prospectFirstName, prospectEmail, prospectCompany, isEnterprise, callName, transcript = [] }) {
  const parties = callData.parties || [];
  const content = callData.content || {};
  const transcriptText = buildTranscriptText(transcript, parties);
  console.info(`[Sales] Transcript: ${transcript.length} sentences, ${transcriptText.length} chars`);

  const primary = primaryInternalSpeaker(transcript, parties);
  const senderName = primary?.name || repName || "the SiteZeus team";

  const aiContent = await extractSalesContent({ transcriptText, prospectFirstName, prospectCompany, senderName });

  let body = null;
  let nextSteps = [];
  let resources = [];

  if (aiContent) {
    body = stripLeadingGreeting((aiContent.body || "").trim()) || null;
    nextSteps = Array.isArray(aiContent.nextSteps) ? aiContent.nextSteps.map(s => String(s).trim()).filter(Boolean) : [];
    resources = resolveResources(aiContent.resources);
  } else {
    // Gong fallback
    body = typeof content.brief === "string" ? content.brief.trim() : null;
    const highlights = content.highlights || [];
    const NEXT_MEETING_TYPES = new Set(["next_meeting", "follow_up", "follow-up", "upcoming_meeting"]);
    const meeting = highlights.find(h => NEXT_MEETING_TYPES.has((h.type || "").toLowerCase()))?.text?.trim();
    if (meeting) nextSteps = [meeting];
    const trackers = content.trackers || [];
    const topics = trackers.filter(t => (t.count || 0) > 0).map(t => (t.name || "").toLowerCase());
    resources = selectResourcesByTopic(topics, isEnterprise);
  }

  const prospectHtml = buildProspectHtml({ prospectFirstName, senderName, body, nextSteps, resources });
  const segmentLabel = isEnterprise ? "Enterprise" : "SMB / Emerging";
  const wrapperHtml = buildRepWrapper({ repEmail, prospectFirstName, prospectEmail, segmentLabel, callName, prospectHtml });
  const subject = `[REVIEW & SEND] Follow-up draft for ${prospectFirstName} — ${callName}`;
  return { subject, wrapperHtml, prospectHtml };
}

function buildProspectHtml({ prospectFirstName, senderName, body, nextSteps, resources }) {
  const parts = [
    `<p>Hi ${escapeHtml(prospectFirstName)},</p>`,
    `<p>Thank you for hopping on the call today.</p>`,
  ];

  if (body) {
    for (const para of body.split(/\n{2,}/).map(p => p.trim()).filter(Boolean)) {
      parts.push(`<p>${escapeHtml(para)}</p>`);
    }
  }

  if (nextSteps.length) {
    parts.push(`<p><strong>Next Steps</strong></p>`);
    parts.push(`<ul>${nextSteps.map(s => `<li>${escapeHtml(s)}</li>`).join("")}</ul>`);
  }

  if (resources.length) {
    parts.push(`<p><strong>Relevant Resources</strong></p>`);
    parts.push(`<ul>${resources.map(r => {
      const link = `<a href="${escapeHtml(r.url)}" style="color:#1a73e8;">${escapeHtml(r.name)}</a>`;
      return r.reason ? `<li>${link} — ${escapeHtml(r.reason)}</li>` : `<li>${link}</li>`;
    }).join("")}</ul>`);
  }

  parts.push(`<p>Thanks,<br>${escapeHtml(senderName)}</p>`);
  return parts.join("\n");
}

function buildRepWrapper({ repEmail, prospectFirstName, prospectEmail, segmentLabel, callName, prospectHtml }) {
  return `
<div style="font-family:Arial,Helvetica,sans-serif;font-size:14px;color:#333;max-width:680px;">
<div style="background:#fff8e1;padding:14px 18px;border-left:5px solid #f9a825;margin-bottom:20px;border-radius:3px;">
<strong>ACTION REQUIRED — Review &amp; Send</strong><br><br>
Draft follow-up for <strong>${escapeHtml(prospectFirstName)}</strong> (<a href="mailto:${escapeHtml(prospectEmail)}">${escapeHtml(prospectEmail)}</a>).<br>
<strong>Segment:</strong> ${escapeHtml(segmentLabel)} &nbsp;|&nbsp; <strong>Call:</strong> ${escapeHtml(callName)}<br><br>
Review below, edit as needed, then send from <strong>${escapeHtml(repEmail)}</strong>.<br>
<em>Do not reply to this message — it was generated automatically.</em>
</div>
<hr style="border:none;border-top:1px solid #ddd;margin:0 0 20px 0;">
<p style="color:#888;font-size:12px;margin-bottom:4px;">Suggested subject: <strong>Great connecting — SiteZeus resources &amp; next steps</strong></p>
<div style="border:1px solid #e0e0e0;padding:20px;border-radius:4px;background:#fafafa;">
${prospectHtml}
</div>
</div>
`;
}
